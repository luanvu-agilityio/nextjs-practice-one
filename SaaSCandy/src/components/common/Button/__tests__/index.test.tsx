import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Button } from '../index';

// ============================================
// SNAPSHOT TEST SUITE
// ============================================
describe('Button Component - Snapshots', () => {
  it('renders all variants correctly', () => {
    const { container: primary } = render(
      <Button variant='primary'>Primary</Button>
    );
    expect(primary).toMatchSnapshot();

    const { container: secondary } = render(
      <Button variant='secondary'>Secondary</Button>
    );
    expect(secondary).toMatchSnapshot();

    const { container: tertiary } = render(
      <Button variant='tertiary'>Tertiary</Button>
    );
    expect(tertiary).toMatchSnapshot();
  });

  it('renders all sizes correctly', () => {
    const { container: small } = render(<Button size='small'>Small</Button>);
    expect(small).toMatchSnapshot();

    const { container: large } = render(<Button size='large'>Large</Button>);
    expect(large).toMatchSnapshot();
  });

  it('renders with default props', () => {
    const { container } = render(<Button>Default Button</Button>);
    expect(container).toMatchSnapshot();
  });

  it('renders disabled state', () => {
    const { container } = render(<Button disabled>Disabled</Button>);
    expect(container).toMatchSnapshot();
  });

  it('renders with custom className', () => {
    const { container } = render(
      <Button className='custom-class'>Custom</Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with complex children', () => {
    const { container } = render(
      <Button>
        <span>Icon</span>
        <span>Text</span>
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders combined variants and sizes', () => {
    const { container } = render(
      <Button variant='secondary' size='small'>
        Combined
      </Button>
    );
    expect(container).toMatchSnapshot();
  });
});

// ============================================
// INTERACTIVE TEST SUITE
// ============================================
describe('Button Component - Interactive', () => {
  it('handles click events', async () => {
    const handleClick = jest.fn();
    const user = userEvent.setup();

    render(<Button onClick={handleClick}>Click me</Button>);
    const button = screen.getByRole('button');

    await user.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('handles multiple clicks', async () => {
    const handleClick = jest.fn();
    const user = userEvent.setup();

    render(<Button onClick={handleClick}>Click me</Button>);
    const button = screen.getByRole('button');

    await user.click(button);
    await user.click(button);
    await user.click(button);

    expect(handleClick).toHaveBeenCalledTimes(3);
  });

  it('does not trigger click when disabled', async () => {
    const handleClick = jest.fn();
    const user = userEvent.setup();

    render(
      <Button onClick={handleClick} disabled>
        Disabled
      </Button>
    );
    const button = screen.getByRole('button');

    await user.click(button);
    expect(handleClick).not.toHaveBeenCalled();
  });

  it('applies aria attributes for accessibility', () => {
    render(
      <Button aria-label='Custom label' aria-pressed='true'>
        Button
      </Button>
    );
    const button = screen.getByRole('button');

    expect(button).toHaveAttribute('aria-label', 'Custom label');
    expect(button).toHaveAttribute('aria-pressed', 'true');
  });

  it('receives focus on tab navigation', async () => {
    const user = userEvent.setup();

    render(<Button>Focusable</Button>);
    const button = screen.getByRole('button');

    await user.tab();
    expect(button).toHaveFocus();
  });

  it('can be activated with keyboard (Enter)', async () => {
    const handleClick = jest.fn();
    const user = userEvent.setup();

    render(<Button onClick={handleClick}>Keyboard</Button>);
    const button = screen.getByRole('button');

    button.focus();
    await user.keyboard('{Enter}');

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('can be activated with keyboard (Space)', async () => {
    const handleClick = jest.fn();
    const user = userEvent.setup();

    render(<Button onClick={handleClick}>Keyboard</Button>);
    const button = screen.getByRole('button');

    button.focus();
    await user.keyboard(' ');

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('passes additional props to button element', () => {
    render(
      <Button data-testid='custom-button' id='btn-1'>
        Props
      </Button>
    );
    const button = screen.getByRole('button');

    expect(button).toHaveAttribute('data-testid', 'custom-button');
    expect(button).toHaveAttribute('id', 'btn-1');
  });
});
