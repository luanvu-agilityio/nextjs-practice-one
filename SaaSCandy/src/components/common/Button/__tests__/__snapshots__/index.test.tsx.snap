// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button Component - Snapshots renders all sizes correctly 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center font-semibold text-md leading-none tracking-normal rounded-xl transition-colors duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 disabled:cursor-not-allowed cursor-pointer bg-orange-background text-white hover:opacity-90 focus-visible:ring-primary px-4 py-2.5"
  >
    Small
  </button>
</div>
`;

exports[`Button Component - Snapshots renders all sizes correctly 2`] = `
<div>
  <button
    class="inline-flex items-center justify-center font-semibold text-md leading-none tracking-normal rounded-xl transition-colors duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 disabled:cursor-not-allowed cursor-pointer bg-orange-background text-white hover:opacity-90 focus-visible:ring-primary px-9 py-5"
  >
    Large
  </button>
</div>
`;

exports[`Button Component - Snapshots renders all variants correctly 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center font-semibold text-md leading-none tracking-normal rounded-xl transition-colors duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 disabled:cursor-not-allowed cursor-pointer bg-orange-background text-white hover:opacity-90 focus-visible:ring-primary px-9 py-5"
  >
    Primary
  </button>
</div>
`;

exports[`Button Component - Snapshots renders all variants correctly 2`] = `
<div>
  <button
    class="inline-flex items-center justify-center font-semibold text-md leading-none tracking-normal rounded-xl transition-colors duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 disabled:cursor-not-allowed cursor-pointer bg-white text-orange-background border border-orange-background hover:bg-orange-background hover:text-white focus-visible:ring-primary px-9 py-5"
  >
    Secondary
  </button>
</div>
`;

exports[`Button Component - Snapshots renders all variants correctly 3`] = `
<div>
  <button
    class="inline-flex items-center justify-center font-semibold text-md leading-none tracking-normal rounded-xl transition-colors duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 disabled:cursor-not-allowed cursor-pointer bg-white text-orange-background border-none hover:text-primary px-9 py-5"
  >
    Tertiary
  </button>
</div>
`;

exports[`Button Component - Snapshots renders combined variants and sizes 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center font-semibold text-md leading-none tracking-normal rounded-xl transition-colors duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 disabled:cursor-not-allowed cursor-pointer bg-white text-orange-background border border-orange-background hover:bg-orange-background hover:text-white focus-visible:ring-primary px-4 py-2.5"
  >
    Combined
  </button>
</div>
`;

exports[`Button Component - Snapshots renders disabled state 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center font-semibold text-md leading-none tracking-normal rounded-xl transition-colors duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 disabled:cursor-not-allowed cursor-pointer bg-orange-background text-white hover:opacity-90 focus-visible:ring-primary px-9 py-5"
    disabled=""
  >
    Disabled
  </button>
</div>
`;

exports[`Button Component - Snapshots renders with complex children 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center font-semibold text-md leading-none tracking-normal rounded-xl transition-colors duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 disabled:cursor-not-allowed cursor-pointer bg-orange-background text-white hover:opacity-90 focus-visible:ring-primary px-9 py-5"
  >
    <span>
      Icon
    </span>
    <span>
      Text
    </span>
  </button>
</div>
`;

exports[`Button Component - Snapshots renders with custom className 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center font-semibold text-md leading-none tracking-normal rounded-xl transition-colors duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 disabled:cursor-not-allowed cursor-pointer bg-orange-background text-white hover:opacity-90 focus-visible:ring-primary px-9 py-5 custom-class"
  >
    Custom
  </button>
</div>
`;

exports[`Button Component - Snapshots renders with default props 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center font-semibold text-md leading-none tracking-normal rounded-xl transition-colors duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 disabled:cursor-not-allowed cursor-pointer bg-orange-background text-white hover:opacity-90 focus-visible:ring-primary px-9 py-5"
  >
    Default Button
  </button>
</div>
`;
